lib/util/close.js

```diff
@@ -61,22 +66,75 @@ exports.cleanModulesDirectory = co.wrap(function *(root, name) {
  * @async
  * @param {NodeGit.Repository} repo
  * @param {String}             submoduleName
+ * @param {Boolean}            force
  */
-exports.close = co.wrap(function *(repo, submoduleName) {
+exports.close = co.wrap(function *(repo, submoduleName, force) {
 
     // This operation is a major pain, first because libgit2 does not provide
     // any direct methods to do the equivalent of 'git deinit', and second
     // because nodegit does not expose the method that libgit2 does provide to
     // delete an entry from the config file.
 
-    // De-initting a submodule requires two things:
-    // 1. Remove all files under the path of the submodule, but not the
+    // De-initting a submodule requires the following things:
+    // 1. Confirms there are no unpushed (to any remote) commits
+    //    or uncommited changes (including new files).
+    // 2. Remove all files under the path of the submodule, but not the
     //    directory itself, which would look to Git as if we were trying
     //    to remove the submodule.
-    // 2. Remove the entry for the submodule from the '.git/config' file.
-    // 3. Remove the directory .git/modules/<submodule>
+    // 3. Remove the entry for the submodule from the '.git/config' file.
+    // 4. Remove the directory .git/modules/<submodule>
+
+    // Confirm there are no unpushed commits and uncommited changes
+
+    if (!force) {
+
+        const submodule = yield NodeGit.Submodule.lookup(repo, submoduleName);
+        const subRepo = yield submodule.open();
+
+        const errorMessages = [];
+
+        // Determine if all of the commits have been pushed to some remote
+
+        let allPushed = false;
+
+        const headCommit = yield subRepo.getHeadCommit();
+        const remotes = yield NodeGit.Remote.list(subRepo);
+
+        const checkUnpushedCommits = co.wrap(function *(remote) {
+            const unpushed = yield GitUtil.listUnpushedCommits(
```

Comment:
The decisions we make based on unpushed commits (which autocorrect wanted
to change to "unpunished") are best-guess; there are ways to circumvent it.
Generally speaking, we're looking for commits that haven't been pushed from
your repository; if they got there some other way, we don't expect to know
about them.

The second point you bring up: "pushing to something that is not under
refs/heads ... There will be no local record of the push." is expected
behavior. `refs/heads` is the standard location for branches to live and
other git-based tools will also be unable to detect pushes to targets
outside `heads`. I think workflows involving non `heads` targets should be
clear that commits pushed to such locations may trigger actions but don't
actually live in the repository in a first-class way.

On Tue, May 24, 2016 at 12:11 PM, Adam Bliss <notifications@github.com>
wrote:

> In lib/util/close.js
> <https://github.com/twosigma/git-meta/pull/36#discussion_r64423060>:
>
> > +    if (!force) {
> > +
> > +        const submodule = yield NodeGit.Submodule.lookup(repo, submoduleName);
> > +        const subRepo = yield submodule.open();
> > +
> > +        const errorMessages = [];
> > +
> > +        // Determine if all of the commits have been pushed to some remote
> > +
> > +        let allPushed = false;
> > +
> > +        const headCommit = yield subRepo.getHeadCommit();
> > +        const remotes = yield NodeGit.Remote.list(subRepo);
> > +
> > +        const checkUnpushedCommits = co.wrap(function *(remote) {
> > +            const unpushed = yield GitUtil.listUnpushedCommits(
>
> That only works if a "git fetch origin" *would* fetch that commit. I.e.,
> the remote ref must be listed in remote.origin.fetch.
>
> If you set remote.origin.fetch to be
> +refs/heads/master:refs/remotes/origin/master rather than the default which
> is usually +refs/heads/*:refs/remotes/origin/* , then you'll find that
> there is no local record of the push.
>
> Alternately, with your current config, try pushing to something that is
> not under refs/heads -- for example, git push HEAD:refs/foo/bar . There
> will be no local record of the push.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/twosigma/git-meta/pull/36/files/df4da8016e5ce8f7ea3c6fda864fa68125ebf783#r64423060>
>